language/compiler/ir-to-bytecode/syntax/src/syntax.rs

```diff
@@ -935,6 +957,30 @@ fn parse_cmd_(tokens: &mut Lexer) -> Result<Cmd_, ParseError<Loc, anyhow::Error>
             let (name, tys) = parse_name_and_type_actuals(tokens)?;
             parse_unpack_(tokens, name, tys)
         }
+        Tok::If => {
```

Comment:
Here are some ramblings, you should probably just need this TLDR:
TLDR
- To me "branch" isn't a verb and it always irks me a bit
- I wanted something that felt more structured. Having `br_true lbl; br_false lbl;` as an if/else felt very akward to me and not the right level for this "statement-like" syntax

--- 
Promised Ramblings: 

The goal here is to have some semblance of readability in that this isn't really the bytecode but just a step up. What I am trying to say is that there will be eventually a version of the IR that is just the bytecode (the mythical `bytecode fun`s) and those will have `Branch`, `BranchTrue` and `BranchFalse`.

And those names are only happening because thats what they are in the bytecode. I don't really like "branch" as a verb and it always reads awkwardly to me personally (this post is sponsored by the x86-jump gang)

I kind of like `if` but maybe I'm insane and my preference is mostly coming from the compiler (where `if` gets compiled to something that is `if (cond) branch else lbl2`)

Reply:
> I still find this command confusing.

+1 to this, and I think your entire critique of the syntax is spot-on. I like your "aggressive" ideas to eliminate some of the frills around the bytecode instructions, and I would also prefer `goto` + `br_true` + `br_false`. Now that you mention it: the bytecode doesn't have a `cond_br <cond> <true_label>, <false_label>` instruction, so it seems unnecessary to me to keep an `if <cond> <true_label> else <false_label>` statement in Move IR.

> By extension, I guess one thing we really need to clarify is, what level of abstraction do we want in the IR?

I began this work with the idea that I'd like to eliminate the Move IR language entirely, essentially replacing it with a machine-parseable textual representation of bytecode instructions. What I began to learn was that the IR language was necessary to make it not only possible, but _easy_, to write bytecode verifier tests. I find that, without having a ton of experience writing and maintaining these IR tests, it's difficult for me to get right the balancing act of "straightforward representation of the underlying bytecode" vs "easy to write and read." So, while I am supportive of the idea of replacing assignment command syntax with `pop`, I have a feeling it will make IR test writing too painful -- what do you think?

Reply:
> What I began to learn was that the IR language was necessary to make it not only possible, but easy, to write bytecode verifier tests. I find that, without having a ton of experience writing and maintaining these IR tests, it's difficult for me to get right the balancing act of "straightforward representation of the underlying bytecode" vs "easy to write and read."

My headspace for this is that the IR should read like psuedocode for bytecode. So high-level but without losing precision. In contrast, the source language is high-level but loses a lot of precision to be more ergonomic/familiar

Reply:
> I kind of like if but maybe I'm insane and my preference is mostly coming from the compiler (where if gets compiled to something that is if (cond) branch else lbl2)

Sorry, not sure I understand what you mean here. Is "branch" an expression or another label?

It's also interesting that you seem to like the if construction from the compiler's perspective. I thought it would be the opposite. Maybe I missed something special about expression-based languages? BTW I had this in my mind when thinking about bytecode generation for if, which made me think that the current if IR command would be unnatural for the compiler to handle.
```
// if (cond) { do_something }

    goto end_if if not (cond);
    do something;
label end_if:
```

```
// if (cond) { do_something } else { do_something_else }

    goto else if not (cond);
    do something;
    goto end_if;

label else:
    do something else;

label end_if:
```

Reply:
By compiler bits I mean literally our source language compiler, not as a general compiler statement. And I think you are thinking a bit too linearly, by that I mean assuming a certain code layout. (Granted the textual code is linear in nature)

But in the source langauge compiler, in the CFG pass, all blocks end in either abort, return, or `if (cond) label else label`, as there is no fall through as there is no implied linearity. 

That is my hope for the IR here is to get to that same model, but it feels a bit weird to absolutely force no fall throughs because the textual format _is_ linear. 
Perhaps this model of thinking about labels+blocks would clear things up a bit in the IR representation

Reply:
> What I am trying to say is that there will be eventually a version of the IR that is just the bytecode (the mythical bytecode funs) and those will have `Branch`, `BranchTrue` and `BranchFalse`. And those names are only happening because thats what they are in the bytecode. I don't really like "branch" as a verb and it always reads awkwardly to me personally (this post is sponsored by the x86-jump gang)

I don't think we ought to spend time renaming the instructions, so I hesitate to mention this, but: @vgao1996 did [write that he'd be fine with changing the bytecode names](https://github.com/diem/diem/pull/9218#discussion_r743857994)... again, I don't think it would be worthwhile to do so, but my interest is primarily in moving things along, so I'd approve a PR that changed the names, or change the underlying bytecode names myself if it helped simplify this discussion a bit. (I don't mean to trivialize the naming aspect of things -- if `branch` was inaccurate to the point of being confusing, then I'd agree that it should be changed... but [it does appear to be an intransitive verb](https://www.wordnik.com/words/branch), and it makes sense to me here, as in "the program's control flow **branches**.")

Reply:
> this post is sponsored by the x86-jump gang

By the way, if this is in reference to some sort of treatise on why x86 named the instruction "jump," then please send a link! Maybe that will get us on the same page, and my web searching hasn't gotten me any good results.

Reply:
I think the new syntax is one of the main ideas in this pull request that isn't quite resolved. Currently I think @vgao1996 and I agree that `if`/`else` syntax is confusing, and something closer to the bytecode instructions would be ideal. I would like to propose the following:

| Proposed Move IR syntax | Move IR syntax in this current PR | Bytecode equivalent | Meaning |
| - | - | - | - |
| `branch b1;` | `goto b1;` | `Branch(<b1-offset>);` | Jump unconditionally to label `b1`. |
| `branch_true(cond) b1;` | Not currently implemented. | `BrTrue(<b1-offset>);` | If `cond` is true jump to `b1`, otherwise fallthrough. |
| `branch_false(cond) b1;` | `if (cond) _ else b1;` | `BrFalse(<b1-offset>);` | If `cond` is true fallthrough, otherwise jump to `b1`. |
| Will not implement. | `if (cond) b1 else b2;` | `BrTrue(<b1-offset>); Branch(<b2-offset>);` | If `cond` is true jump to `b1`, otherwise jump to `b2`. |

My rationale for the above being that the proposed syntax mirrors the bytecode, and eliminates syntax (currently in this PR) that translates to 2 separate bytecode instructions: `if (cond) b1 else b2;`. Thoughts @vgao1996 and @tnowacki?